README
===========================================================
HLstatsX Community Edition
http://www.hlxce.com
===========================================================

HLstatsX Community Edition is an open-source project licensed
under GNU General Public License v2 and is a real-time stats
and ranking for Source engine based games. HLstatsX Community
Edition uses a Perl daemon to parse the log streamed from the
game server. The data is stored in a MySQL Database and has
a PHP frontend.


For installation help, please visit our web site:

	http://www.hlxce.com
	http://wiki.hlxce.com
	
or our forums:

	http://forums.hlxce.com
	
or join us on IRC:

	irc.gamesurge.net, channel #HLXCE
	

To be notified of upcoming releases, you can join our
FreeLists group:

	http://www.freelists.org/list/hlxce
	
	
We are also always looking for Beta-Testers.  Join our
Google Group at:

	http://groups.google.com/group/hlxce-beta-testers
	

NOTE:

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

CHANGELOG
HLstatsX Community Edition 1.6.19
October 13, 2012

+ Heatmaps: Added support for NTS, L4D2 and CSS Heatmap configurations. (#1602, #844)
! GameSupport: Fixed some CS:GO actions that were incorrectly added in 1.6.18. (#1599)
! Web: Fixed adding a new server to a duplicate game returning user to original game's server listing. (#1594)
! Web: Fixed output that could be used as Cross-Site Scripting exploit (Reported by P.Whitton) (#1612)

HLstatsX Community Edition 1.6.18
August 31, 2012

+ GameSupport: Added support for Dino D-Day and Counter-Strike: Global Offensive. (#1589, #1552)
+ Web: Added additional chat logging support for Nuclear Dawn.
! Web: Fixed adding a new server to a duplicate game returning user to original game's server listing. (#1594)


HLstatsX Community Edition 1.6.17
June 27, 2012

+ GameSupport: Added 9 new TF2 weapons from the Pyromania update. (#1579, 1583)
+ Web: Added new map image for TF2 map sd_doomsday. (#1582)
! GameSupport: Corrected TF2 award names that were incorrectly updated. (#1558)
! Web: Updated updated check URL to hlxce.com domain. (#1581)


HLstatsX Community Edition 1.6.16
May 20, 2012

+ Daemon: Added a task to purge out old entries in hlstats_server_load table. (#1545)
+ GameSupport: Added 8 new weapons for TF2. (#1546 #1547)
+ GameSupport: Support for Nuclear Dawn added. (#1487)
    SuperLogs is available for Nuclear Dawn.  More info at http://wiki.hlxce.com/wiki/SuperLogs
! Daemon: Fixed awards being given to players of different games. (#1440)
! Daemon: Fixed daemon awarding no points when you kill a player with no points. (#1227)
! Daemon: Fixed regression causing the 'next' command to have garbage output. (#1479)
! Daemon: Fixed uptime regex calculation due to format change in Team Fortress 2. (#1497)
! Daemon: Updated geoip.py to use correct MaxMind URL for GeoLiteCity.CSV file. (#1521)
! Daemon: Fixed install_binary.sh not installing GeoLiteCity to correct location. (#1532)
! General: Standardized typing for server_id columns in all tables. (#1456)
! Web: Removed legacy which could be used for XSS and SQL injection attacks.
! Web: Fixed Status and Avatar not working for players with custom URLs in their Steam profile. (#1525)
! Web: Fixed various locations where role code was displayed instead of role name. (#1548)
! Web: Added missing TF2 award and ribbon images. (#1449 #1468)
! Web: Fixed a syntax issue on tools_perlcontrol.php that caused the page to not load on PHP 5.4 (#1544)
? Web: Changed the default zoom level for the EUROPE region by 1.  (#1426)
? Web: Changed the table on player award history page to use the award name instead of award code. (#1556)
? Web: Changed the query on the Chat page to sort by the ID instead of Event Time, which would have required an index.  (#1083)


HLstatsX Community Edition 1.6.15
November 8th, 2011

+ GameSupport: Added many new weapons for TF2, including the Victory Pack, Manno-Technology and Halloween update (#1439, #1447, #1448) (Packhead)
+ GameSupport: Added awards and ribbons for four existing actions (#1467) (Packhead)
+ GameSupport: Captured player_penetration weapon event and changed to action for TF2 (weapon logged as machina) (Packhead)
+ Web: Awards, Maps, Ribbons, Roles and Weapon images now failback to "realgame" directory if not present in "game" directory (#1461) (Packhead)
! Daemon: Fixed FPS and Uptime tracking post-TF2 stats output fix (#1458) (Packhead)
! Daemon: Converted server load DB updates to cached queries to reduce load (Packhead)
! Daemon: Converted almost all daemon queries to use MySQL Prepared Statements to help reduce DB load (Packhead)
! GameSupport: Corrected typos and added two missing TF2 weapons (#1446) (Bluthund)
? Plugin: Removed old hack to fix cvars on a2s_rules response (psychonic)
? Plugin: Removed a large number of redundant client checks (psychonic)
? Web: Removed unnecessary joins from multiple pages (octo-dhd)
? Web: Additional index added to speed up PlayerPlayerAction transactions (octo-dhd)
? Web: Updater has been enhanced to provide more output (Packhead)


HLstatsX Community Edition 1.6.14
July 15, 2011
+ GameSupport: Added 31 new weapons for Team Fortress 2 (#1421, #1429) (Packhead, psychonic, octo-dhd, bug, Rakshot, soolshock, solarpowered)
+ GameSupport: Added new TF2 player action to log scout extinguishes with the Mad Milk (#1397) (Packhead)
! Daemon: Player position was not printing at initial release of 1.6.13 (#1425) (psychonic)
! Daemon: Updated the GeoLite_Import.sh script due to change in release cycle of Max Mind (#1438) (aubergine)
? Daemon: Chat commands are no longer reliant on player events being enabled (psychonic)
? Daemon: Lazy log line format no longer clears player's team (psychonic)
? Daemon: Removed some redundant logic in EventHandlers.plib (psychonic)
? GameSupport: Renamed 14 Team Fortress 2 weapons to align them with the TF2 wiki (#1423) (Packhead)


HLstatsX Community Edition 1.6.13
January 22, 2011
+ GameSupport: Added missing headshot action, awards, and ribbons to HL2DM (psychonic)
+ General: Added total teamkills to player tracking (psychonic)
+ General: Added teamkills and kill/death streaks to player session tracking (psychonic)
+ General: Added more new special award possibilities, "connectiontime", "killsteak", and "deathstreak" (psychonic)
+ Web: Added teamkills to ingame statsme webpage (psychonic)
? GameSupport: Updated many recently-changed TF2 weapon codes (psychonic)
? Daemon: Updated run_hlstats to now automatically change to the directory it's in (no longer do you have to CD before using script in cron!) (Packhead)
? Daemon: Updated run_hlstats to verify shebang line in hlstats.pl is formatted correctly and binary exists (Packhead)
? Daemon: Moved connect message from Connect event to EnterGame event (to prevent display when banned users connect) (psychonic)
? Heatmaps: Greatly improved output quality of heat data (msleeper)

web/hlstats.php
<?php
/*
HLstatsX Community Edition - Real-time player and clan rankings and statistics
Copyleft (L) 2008-20XX Nicholas Hastings (nshastings@gmail.com)
http://www.hlxcommunity.com

HLstatsX Community Edition is a continuation of 
ELstatsNEO - Real-time player and clan rankings and statistics
Copyleft (L) 2008-20XX Malte Bayer (steam@neo-soft.org)
http://ovrsized.neo-soft.org/

ELstatsNEO is an very improved & enhanced - so called Ultra-Humongus Edition of HLstatsX
HLstatsX - Real-time player and clan rankings and statistics for Half-Life 2
http://www.hlstatsx.com/
Copyright (C) 2005-2007 Tobias Oetzel (Tobi@hlstatsx.com)

HLstatsX is an enhanced version of HLstats made by Simon Garner
HLstats - Real-time player and clan rankings and statistics for Half-Life
http://sourceforge.net/projects/hlstats/
Copyright (C) 2001  Simon Garner
            
This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.

For support and installation notes visit http://www.hlxcommunity.com
*/

define('IN_HLSTATS', true);
require('config.php');
$historical_cache=0;
if(defined('HISTORICAL_CACHE'))
{
	$historical_cache=constant('HISTORICAL_CACHE');
}

if($historical_cache==1)
{
	$rawmd5=md5(http_build_query($_REQUEST));
	$dir1=substr($rawmd5,0,1);
	$dir2=substr($rawmd5,1,1);
	$cachetarget=sprintf("cache/%s/%s/%s", $dir1, $dir2, $rawmd5);

	@mkdir("cache/$dir1");
	@mkdir("cache/$dir1/$dir2");

	if(file_exists($cachetarget))
	{
		file_put_contents("cache/cachehit",$cachetarget . "\n", FILE_APPEND);
		echo file_get_contents($cachetarget);
		die;
	}
}

session_start();

if((!empty($_GET['logout'])) && $_GET['logout'] == '1') {
        unset($_SESSION['loggedin']);
        header("Location: http://".$_SERVER['HTTP_HOST'].$_SERVER['SCRIPT_NAME']);
        die;
}

// Several stuff added by Malte Bayer
global $scripttime, $siteurlneo;
$scripttime = microtime(true);
$siteurlneo='http://'.$_SERVER['HTTP_HOST'].substr($_SERVER['PHP_SELF'],0,strpos($_SERVER['PHP_SELF'],strrchr($_SERVER['PHP_SELF'],'/'))+1);
$siteurlneo=str_replace('\\','/',$siteurlneo);

// Several Stuff end

foreach ($_SERVER as $key => $entry) {
	if ($key !== 'HTTP_COOKIE') {
		$search_pattern  = array('/<script>/', '/<\/script>/', '/[^A-Za-z0-9.\-\/=:;_?#&~]/');
		$replace_pattern = array('', '', '');
		$entry = preg_replace($search_pattern, $replace_pattern, $entry);
  
		if ($key == "PHP_SELF") {
			if ((strrchr($entry, '/') !== '/hlstats.php') &&
				(strrchr($entry, '/') !== '/ingame.php') &&
				(strrchr($entry, '/') !== '/show_graph.php') &&
				(strrchr($entry, '/') !== '/sig.php') &&
				(strrchr($entry, '/') !== '/sig2.php') &&
				(strrchr($entry, '/') !== '/index.php') &&
				(strrchr($entry, '/') !== '/status.php') &&
				(strrchr($entry, '/') !== '/top10.php') &&
				(strrchr($entry, '/') !== '/config.php') &&
				(strrchr($entry, '/') !== '/') &&
				($entry !== '')) {
				header("Location: http://$siteurlneo/hlstats.php");    
				exit;
			}    
		}
		$_SERVER[$key] = $entry;
	}
}

@header('Content-Type: text/html; charset=utf-8');

// do not report NOTICE warnings
@error_reporting(E_ALL ^ E_NOTICE);

////
//// Initialisation
////

define('PAGE', 'HLSTATS');

///
/// Classes
///

// Load required files
require(INCLUDE_PATH . '/class_db.php');
require(INCLUDE_PATH . '/class_table.php');
require(INCLUDE_PATH . '/functions.php');

$db_classname = 'DB_' . DB_TYPE;
if ( class_exists($db_classname) )
{
	$db = new $db_classname(DB_ADDR, DB_USER, DB_PASS, DB_NAME, DB_PCONNECT);
}
else
{
	error('Database class does not exist.  Please check your config.php file for DB_TYPE');
}

$g_options = getOptions();

if (!isset($g_options['scripturl'])) {
	$g_options['scripturl'] = isset($_SERVER['PHP_SELF']) ? $_SERVER['PHP_SELF'] : getenv('PHP_SELF');
}

////
//// Main
////

$game = valid_request(isset($_GET['game'])?$_GET['game']:'', 0);

if (!$game)
{
	$game = isset($_SESSION['game'])?$_SESSION['game']:'';
}
else
{
	$_SESSION['game'] = $game;
}

if (!$realgame && $game)
{
	$realgame = getRealGame($game);
	$_SESSION['realgame'] = $realgame;
}

$mode = isset($_GET['mode'])?$_GET['mode']:'';

$valid_modes = array(
	'players',
	'clans',
	'weapons',
	'roles',
	'rolesinfo',
	'maps',
	'actions',
	'claninfo',
	'playerinfo',
	'weaponinfo',
	'mapinfo',
	'actioninfo',
	'playerhistory',
	'playersessions',
	'playerawards',
	'search',
	'admin',
	'help',
	'bans',
	'servers',
	'chathistory',
	'ranks',
	'rankinfo',
	'ribbons',
	'ribboninfo',
	'chat',
	'globalawards',
	'awards',
	'dailyawardinfo',
	'countryclans',
	'countryclansinfo',
	'teamspeak',
	'ventrilo',
	'updater',
	'profile'
);
   
if (file_exists('./updater') && $mode != 'updater')
{
	pageHeader(array('Update Notice'), array('Update Notice' => ''));
	echo "<div class=\"warning\">\n" . 
	"<span class=\"warning-heading\"><img src=\"".IMAGE_PATH."/warning.gif\" alt=\"Warning\"> Warning:</span><br />\n" .
	"<span class=\"warning-text\">The updater folder was detected in your web directory.<br />
	To perform a Database Update, please go to <strong><a href=\"{$g_options['scripturl']}?mode=updater\">HLX:CE Database Updater</a></strong> to perform the database update.<br /><br />
	<strong>If you have already performed the database update, <strong>you must delete the \"updater\" folder from your web folder.</span>\n</div>";
	pageFooter();
	die();
}
   
if ( !in_array($mode, $valid_modes) )
{
	$mode = 'contents';
}

if ( file_exists(PAGE_PATH . "/$mode.php") )
{
	@include(PAGE_PATH . "/$mode.php");
	pageFooter();
}
else
{
	header('HTTP/1.1 404 File Not Found', false, 404);
	error('Unable to find ' . PAGE_PATH . "/$mode.php");
	pageFooter();
}

?>
